File: docs/Sword2offer/CH3-高质量代码/Python/reverseLinkedList.py
class Item:
  def __init__(self, value):
    self.value = value
    self.next = None

class LinkedList:
  def __init__(self):
    self.head = None

  def append(self, value):
    if self.head == None:
      self.head = Item(value)
    else:
      item = Item(value)
      current = self.head
      while current.next:
        current = current.next
      current.next = item

  def show(self):
    current = self.head
    while current:
      print(current.value)
      current = current.next
  
  def __recurivse_reverse(self, node):
    if node:
      if node.next:
        next_node = self.__recurivse_reverse(node.next)
        next_node.next = node
      else:
        self.head = node
      return node

  def reverse(self):
    if not self.head:
      return
    tail = self.__recurivse_reverse(self.head)
    tail.next = None


if __name__ == '__main__':
  linkedList = LinkedList()
  linkedList.append(1)
  linkedList.append(2)
  linkedList.append(3)
  linkedList.append(4)
  linkedList.reverse()
  linkedList.show()

File: docs/Sword2offer/CH3-高质量代码/ES6/reverseLinkedList.js
let LinkedList = function () {
  let head = null
  let Item = function (value) {
    this.value = value
    this.next = null
  }
  this.push = (value) => {
    if (!head) {
      head = new Item(value)
    } else {
      let item = new Item(value),
        current = head
      // 找到末尾节点
      while (current.next) {
        current = current.next
      }
      current.next = item
    }
  }
  this.getHead = () => {
    return head
  }
}

let reverse = () => {
  let tail = null
  let reverseLinkedlist = (root) => {
    if (root) {
      if (root.next) {
        let nextNode = reverseLinkedlist(root.next)
        nextNode.next = root
        return root
      } else {
        tail = root
        return root
      }
    }
  }
  return function (root) {
    reverseLinkedlist(root)
    root.next = null
    return tail
  }
}

let linkedList = new LinkedList()
linkedList.push(1)
linkedList.push(2)
linkedList.push(3)
linkedList.push(4)
let reverseor = reverse()
let rLinkedList = reverseor(linkedList.getHead())
while (rLinkedList) {
  console.log(rLinkedList.value)
  rLinkedList = rLinkedList.next
}
